File: 3980-best-time-to-buy-and-sell-stock-using-strategy/best-time-to-buy-and-sell-stock-using-strategy.cpp
class Solution {
public:
    long long maxProfit(vector<int>& prices, vector<int>& s, int k) {
        long long ans = 0;
        int n = prices.size();
        for(int i=0;i<n;i++){
            ans +=1LL*prices[i]*s[i];
        }
        vector<long long>pre(n+1);
        pre[0] = 0;
        for(int i=0;i<n;i++){
            pre[i+1] = pre[i]+prices[i];
        }
        vector<long long>pre2(n+1);
        pre2[0] = 0;
        for(int i=0;i<n;i++){
            pre2[i+1] = pre2[i]+prices[i]*s[i];
        }
        int x = k/2;
        for(int i=0;i<=n-k;i++){
            long long p = pre[i+k] - pre[i+x];
            p += pre2[i]-pre[0];
            p += pre2[n]-pre2[i+k];
            ans = max(ans,p);
        }
        return ans;
    }
};

File: 3980-best-time-to-buy-and-sell-stock-using-strategy/best-time-to-buy-and-sell-stock-using-strategy.java
class Solution {
    public long maxProfit(int[] prices, int[] strategy, int k) {
        int n = prices.length;
        long profitPrefix[] = new long[n];
        long pricePrefix[] = new long[n];
        profitPrefix[0] = (long) prices[0] * strategy[0];
        pricePrefix[0] = prices[0];
        for(int i = 1 ; i < n ; ++i){
            profitPrefix[i] = profitPrefix[i-1] + (long)prices[i]*strategy[i];
            pricePrefix[i] = pricePrefix[i-1]+prices[i];
        }
        long baseProfit = profitPrefix[n-1];
        long res = baseProfit;
        int i = 0 , j = k-1;
        int mid = k / 2;
        while(j < n){
            long oldWindow = i == 0 ? profitPrefix[j] : profitPrefix[j] - profitPrefix[i-1] ;
            long newWindow = mid == 0 ? pricePrefix[j] : pricePrefix[j] - pricePrefix[mid-1];
            long newSum = baseProfit - oldWindow + newWindow;
            res = Math.max(res,newSum);
            ++i;
            ++mid;
            ++j;
        }
        return res;

    }
}